### DISCUSSION TITLE
`User has not set a password` error on change password API call

### DISCUSSION THREAD

**Discussion by ashep** [zitadel/zitadel#7634] [Q&A] [answered]
Hi there!

1. A user is created using [this method](https://zitadel.com/docs/apis/resources/user_service/user-service-add-human-user) with `password.password` set to an initial password and `password.changeRequired` set to `false`.
2. The [change password](https://zitadel.com/docs/apis/resources/user_service/user-service-set-password) method is called with `currentPassword` set to the initial password, `newPassword.password` set to a new password, and `newPassword.changeRequired` set to `false`.
3. The `User has not set a password` error is received.

What am I doing wrong? Thank you.

**Comment by hifabienne** [comment]
Hi @ashep 
I tried to reproduce your problem, but it worked fine for me.
Can you please send the complete requests you are doing, for example as curl?

**Reply by ashep** [reply]
@hifabienne, thank you for the answer. I've just tried to perform queries via cURL and found out that it works, actually. It seems my problem is caused by gRPC API misusage. I am going to figure out what I am doing wrong. Thank you for your support.

**Reply by ashep** [reply]
@hifabienne got it! Error appears when I try to change a user's password from a non-default organization.

Create user request (note the `organization` field is set):

```shell
curl -X POST -H 'Content-Type: application/json' -H 'Authorization: Bearer TOKEN' --data '{"organization": { "orgId": "259979859881623554"}, "profile":{"givenName": "GN", "familyName": "FN"}, "email":{"email":"foo@bar.baz","isVerified": true}, "password":{"password":"theP@ssw0rd123", "changeRequired": false}}' http://zitadel:8081/v2beta/users/human
```

Create user response:

```json
{"userId":"259980164975296514", "details":{"sequence":"2", "changeDate":"2024-03-26T12:30:54.975844Z", "resourceOwner":"259979859881623554"}}
```

Change password request:

```shell
curl -X POST -H 'Content-Type: application/json' -H 'Authorization: Bearer TOKEN' --data '{"currentPassword": "theP@ssw0rd123", "newPassword": {"password": "theNewP@ssw0rd321", "changeRequired": false}}' http://zitadel:8081/v2beta/users/259980164975296514/password
```
Change password response:

```json
{"code":9, "message":"User has not set a password (COMMAND-Fds3s)", "details":[{"@type":"type.googleapis.com/zitadel.v1.ErrorDetail", "id":"COMMAND-Fds3s", "message":"User has not set a password"}]}
```

**Reply by hifabienne** [reply]
I could reproduce this. Let me check internally and get back to you

**Reply by ashep** [reply]
@hifabienne, there is also a related issue. When I use the verification code instead of the current password, and the user is not from the default org, I get the "code not found" error.

**Reply by hifabienne** [reply]
Ok looks like we take the context of the current user into account at the moment, which should not be.
We are already aware of the problem, and working on a fix

**Reply by ashep** [reply]
Hi @hifabienne! Do you happen to have any updates on this?

**Reply by hifabienne** [reply, marked as answer]
Hi @ashep I quickly tested this again, and it looks good

**Reply by ashep** [reply]
Thank you!
